### PR TITLE
(Finally) Python3 codebase...

### PR DESCRIPTION
**Pull request by JedMeister**
This is a python3 compatible version of Octohub that builds to Debian package. There may still be improvements which could be made, but this is working sweet IMO...

FWIW, for TurnKey users, there is a package of this code in the `buster-testing` repo (`octohub=0.3`).

Should close https://github.com/turnkeylinux/tracker/issues/1540

### CODE REVIEW

**Comment by JedMeister** [comment]
@ongle - Can you please review this so we can merge?!

File: contrib/offline-issues/parse.py [hunk 1]
```diff
@@ -75,8 +81,7 @@ def output_issues(issues, outdir):
         path = os.path.join(outdir, 'all', str(issue.number))
         path_symlink = '../../all/%s' % str(issue.number)
         mkdir(os.path.dirname(path))
-        file(path, 'w').write(json.dumps(issue, indent=1))
-
+        open(path, 'w').write(json.dumps(issue, indent=1))
```
Comment (OnGle) [inline contrib/offline-issues/parse.py:84]: Please use context manager here. Whilst probably safe let's not depend on python GC to close our file handles. 

Suggested fix:
```python3
with open(path, 'w') as fob:
    json.dump(issue, fob, indent=1)
```

File: contrib/gist/gist.py [hunk 2]
```diff
@@ -38,29 +38,33 @@
...
-    print >> sys.stderr, __doc__.lstrip()
+    print('Syntax: %s [-options] [file...fileN]' % cmd, file=sys.stderr)
+    print(__doc__.lstrip(), file=sys.stderr)
 
     sys.exit(1)
 
+
 def render_gist(gist):
-    files = ' '.join(gist.files.keys()).encode('ascii', 'ignore').strip()
+    files = ' '.join(list(gist.files.keys())).encode('ascii', 'ignore').strip()
```
Comment (OnGle) [inline contrib/gist/gist.py:63]: Do we need to turn this into a list before joining? Not a big issue but afaik `str.join` takes anything of type `Iterator[str]`

File: contrib/offline-issues/parse.py [hunk 3]
```diff
@@ -120,9 +126,10 @@ def main():
             if os.path.exists(path):
                 shutil.rmtree(path)
 
-    issues_dict = json.loads(file(infile, 'r').read())
+    issues_dict = json.loads(open(infile, 'r').read())
```
Comment (OnGle) [inline contrib/offline-issues/parse.py:129]: Again file handle is only closed by GC, suggested fix:

```python3
with open(infile, 'r') as fob:
    issues_dict = json.load(fob)
```

**Review by OnGle — CHANGES_REQUESTED** [review]
I would like these changes to be made. After that I'm happy to merge

**Comment by OnGle** [comment]
I've got some further changes locally, but they'll take a while to finalize. I suggest we do 2 separate releases, as frustrating as this might be I do have rational.

- I would like to have complete typing for the GitHub API, however typing this complex will likely only be supported 3.7+ possibly later but I don't recall the specifics off the top of my head. This will break any kind of type-inspection performed by dependent libraries which otherwise would only require a python 2->3 upgrade. (Even without complete typing for the GitHub API there are several API breaking changes that will be required to fully implement typing into octohub)
- I will be replacing the ``getopt`` usage with ``argparse``, possibly modifying the CLI interface to some-extent, where this *should* be a drop-in replacement for CLI usage.

TL;DR despite not being python version compatible this should be, for the most part API compatible both for CLI usage and as a library, however my subsequent changes will not.

If you disagree, it may be worthwhile holding off the merge.

File: contrib/offline-issues/parse.py (hunk 1 above)
Comment (JedMeister) [reply contrib/offline-issues/parse.py:84]: Done.

File: contrib/gist/gist.py (hunk 2 above)
Comment (JedMeister) [reply contrib/gist/gist.py:63]: This was an automated 2to3 change; not one I made manually. So I'm assuming that it needs to be done, or is at least good practice to be explicit?!

File: contrib/offline-issues/parse.py (hunk 3 above)
Comment (JedMeister) [reply contrib/offline-issues/parse.py]: Done

**Comment by JedMeister** [comment]
@OnGle - I've addressed your concerns and pushed.

As to your suggestions on typing and move to `argparse`, I agree.

I also think that we should ditch the `Makefile`. I have already moved the (Debian) build to `pybuild`, so the `Makefile` is probably already out of date and I'm not even sure if it still works?!

File: contrib/gist/gist.py (hunk 2 above)
Comment (OnGle) [reply contrib/gist/gist.py:63]: My guess is this is another case of https://stackoverflow.com/questions/17695456/why-does-python-3-need-dict-items-to-be-wrapped-with-list effectively this ``keys`` attribute was a ``List[str]`` in python2, now it's some kind of ``Iterator[str]``, the easiest way ensure behavior doesn't change is to turn it into a list, afaik it doesn't matter here but if we were indexing/slicing or something else which wont work on an iterator but does on a string it would cause issues.

My guess it is neither required nor best practice but simply a safer change for 2to3 to make for the general case.

**Comment by OnGle** [comment]
I've noted I think the ``list(...)`` is still not required inline, but it should make no difference and personally I'd prefer to merge over holding up octohub anymore.

File: contrib/gist/gist.py (hunk 2 above)
Comment (JedMeister) [reply contrib/gist/gist.py:63]: Ah ok, thanks @OnGle. That makes sense. TBH, my inclination is to just leave it as is. But if you would prefer to tidy it up, then I'm open to that. I'll leave it to you...
